docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,447 @@
+  // +optional
+  ValuesTemplate string `json:"values,omitempty"`
+}
+
+// HelmChartProxyStatus defines the observed state of HelmChartProxy.
+type HelmChartProxyStatus struct {
+  // Conditions defines current state of the HelmChartProxy.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+
+  // MatchingClusters is the list of references to Clusters selected by the ClusterSelector.
+  // +optional
+  MatchingClusters []corev1.ObjectReference `json:"matchingClusters"`
+}
+```
+
+#### HelmReleaseProxy
+
+HelmReleaseProxy represent a single Helm release installed on a selected workload Cluster. HelmReleaseProxies are created and managed by the HelmChartProxy controller. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).
+
+```go
+// HelmReleaseProxySpec defines the desired state of HelmReleaseProxy.
+type HelmReleaseProxySpec struct {
+  // ClusterRef is a reference to the Cluster to install the Helm release on.
+  ClusterRef *corev1.ObjectReference `json:"clusterRef"`
+
+  // ChartName is the name of the Helm chart in the repository.
+  ChartName string `json:"chartName"`
+
+  // RepoURL is the URL of the Helm chart repository.
+  RepoURL string `json:"repoURL"`
+
+  // ReleaseName is the release name of the installed Helm chart. If it is not specified, a
+  // name will be generated.
+  // +optional
+  ReleaseName string `json:"releaseName,omitempty"`
+
+  // ReleaseNamespace is the namespace the Helm release will be installed on the referenced 
+  // Cluster. If it is not specified, the default namespace will be used.
+  // +optional
+  ReleaseNamespace string `json:"namespace,omitempty"`
+
+  // Version is the version of the Helm chart. If it is not specified, the chart will use 
+  // and be kept up to date with the latest version.
+  // +optional
+  Version string `json:"version,omitempty"`
+
+  // Values is an inline YAML representing the values for the Helm chart. This YAML is the result of the rendered
+  // Go templating with the values from the referenced workload Cluster.
+  // +optional
+  Values string `json:"values,omitempty"`
+}
+
+// HelmReleaseProxyStatus defines the observed state of HelmReleaseProxy.
+type HelmReleaseProxyStatus struct {
+  // Conditions defines current state of the HelmReleaseProxy.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+
+  // Status is the current status of the Helm release.
+  // +optional
+  Status string `json:"status,omitempty"`
+
+  // Revision is the current revision of the Helm release.
+  // +optional
+  Revision int `json:"revision,omitempty"`
+
+  // Namespace is the namespace of the Helm release on the workload cluster.
+  // +optional
+  Namespace string `json:"namespace,omitempty"`
```

Comment:
The namespace in the spec corresponds to the namespace argument in `helm install --namespace`, which is described as the "namespace scope for this request". The namespace in the status corresponds to the output of `helm status` shown below. I'm not sure if it's guaranteed to be identical but I just based it off of the Helm API.

```
NAME: nginx-ingress-1663364363
LAST DEPLOYED: Fri Sep 16 17:39:26 2022
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
The NGINX Ingress Controller has been installed.
```

Reply:
I would have assumed that it's always the same. Essentially if I want to deploy a chart in a certain namespace I would expect it to end up there :)

Reply:
Fair enough, I can ask on the Helm repo to clarify. I figured they probably had a reason to put it in the status (in case there's some special edge cases).

Reply:
Would be interesting to know

Reply:
I opened an issue on the Helm repo to clarify but I think we can probably assume it's the same. In that case do you want me to just drop the field from the status and reference the field in the spec?

Reply:
I think it makes sense to drop it for now. We can re-add it later if we find out that it is important.

What do you mean with:
> reference the field in the spec

Reply:
Similar to the release name, if the field isn't set, it'll be set to a default or generated value by the webhook/controller.

Reply:
I think it would be good to have a authoritative value in the spec (similar to the release name).

I would either make it mandatory or default it to the same value as the helm cli (which is probably "default")

Reply:
Fair enough, I'll make it required and we can always change it later if there's an additional edge case we didn't consider.

Reply:
One nit. As of today kubectl does client-side OpenAPI validation per default (it can be disabled with --validate=false). The client-side validation will be dropped in the future.

But as of today making a field mandatory in the API type will force most of our users (who don't know about --validate=false or don't want to use it) to set the field on client-side.

So as of today if you want to make it optional for the regular kubectl user you have to keep it optional.

Going forward, I'm not sure how it works but probably the mutating webhooks are called before OpenAPI validation (no guarantees).

Just wanted to provide context, up to you and it's v1alpha1 so we can change it over time.

It probably makes sense though to define both ReleaseName and RleeaseNamespace the same way

Reply:
Actually I'll probably keep the ReleaseNamespace optional and have a defaulting webhook set it in the HelmReleaseProxy if we need to. But yeah I'm open to iterating on how we want to handle fields like this.

Reply:
Sounds good to me